### ISSUE TITLE
Bug in retrieving the apiVersion with convention routing

### ISSUE THREAD

**Issue by dawaji** [Microsoft/aspnet-api-versioning#225]
Hi,

I'm trying to match candidates on actions rather than controllers in the ResolveController and i' m struggling with a strange bug.
I'm  trying to retrieve the versions of an action decorated with the **MapToApiVersion** attribute with the convention Routing.

For that, I 've changed the way of finding the matches in the **ResolveController** method of the **ConventionRouteControllerSelector** class. See below my linq request. I needed the ApiVersionActionSelector to find the actionDescriptors of a controllerDescriptor.

```
...
static ApiVersionActionSelector apiVersionActionSelector = new ApiVersionActionSelector();
...
static HttpControllerDescriptor ResolveController(IEnumerable<HttpControllerDescriptor> candidates, ApiVersion requestedVersion, Lazy<Exception> ambiguousException)
...
 var matches = candidates
                    .Select(candidate => new { ControllerDescriptor = candidate, ActionDescriptors = apiVersionActionSelector.GetActionMapping(candidate).SelectMany(a => a) })
                    .SelectMany(b => b.ActionDescriptors.Select(actionDescriptor => new { b.ControllerDescriptor, ActionDescriptor = actionDescriptor, Versions = actionDescriptor.GetApiVersions() }))
                    .Where(e => e.Versions.Contains(requestedVersion) && e.ControllerDescriptor.GetDeclaredApiVersions().Contains(requestedVersion));

```

Below my example :
```
     [ApiVersion("1.0")]`
    [ApiVersion("2.0")]
    public class PersonController : ApiController
    {
        [MapToApiVersion("1.0")]
        // GET: api/Person
        public IEnumerable<string> Get()
        {
            return new string[] { "V1", "V1" };
        }
    }
```
```
    [ApiVersion("2.0")]
    [ControllerName("Person")]
    public class Person2Controller : ApiController
    {
        [MapToApiVersion("2.0")]
        // GET: api/Person
        public IEnumerable<string> Get()
        {
            return new string[] { "V2", "V2"};
        }
    }
```
If i called http://localhost:52429/api/Person?api-version=2, the system is able to find the action, but if i called http://localhost:52429/api/Person?api-version=1 and after i call http://localhost:52429/api/Person?api-version=2, the system throws a **ambiguousException**.

After investigating, i found that the the method GetApiVersions(), doesn't work correctly, because the action Get of the V1 controller returns 2.0

Thanks.

**Comment by commonsensesoftware** [comment]
The `GetApiVersions()` method is just a shortcut for the _declared_ API versions on an action a la `GetApiVersionModel().DeclaredApiVersions`. The `GetApiVersionModel()` extension method for an action enumerates the attributes on the action and stores it in the associated property bag, which can be found in [ApiVersionModel.cs](../blob/master/src/Microsoft.AspNet.WebApi.Versioning/Versioning/ApiVersionModel.cs#L85).

It would seem you have duplicates or crisscrossed controller/action mapping in your query. While I won't rule out the possibility of a bug, I'd need evidence of this behavior without any of the modifications. The order in which services are called not only shouldn't matter, but will certainly happen out of order.

I'm not sure if I have a test for this exact scenario, but I would be very surprised to find this type of bug after so much time and adoption. It seems to be such an obvious flaw. In fairness though, we call them _bugs_ for a reason and I won't rule out the possibility just yet.

One more vital thing to keep in mind trying to evaluate the API versions. The `GetApiVersionModel` is super useful for service authors, but the complete model is **not** assembled during evaluation. The full ApiVersionModel is not aggregated until a match has been made. Building the complete model for callers higher in the stack are one of the value propositions the framework provides, but also is where there are challenges in the implementation.

I struggled with that part of the implementation myself, which is why there are both unit tests and acceptance tests. I ran into a few issues that only surfaced when running through the entire Web API stack. However, trying to test and troubleshoot basic behaviors purely through the full stack is also quite daunting. I suggest starting with simple unit tests and then moving up to verifying end-to-end scenarios.  You can look at [ApiVersionControllerSelectorTest.cs](../blob/master/test/Microsoft.AspNet.WebApi.Versioning.Tests/Dispatcher/ApiVersionControllerSelectorTest.cs) and [ApiVersionActionSelectorTest.cs](../blob/master/test/Microsoft.AspNet.WebApi.Versioning.Tests/Controllers/ApiVersionActionSelectorTest.cs) for ideas on how to setup new tests.

**Comment by commonsensesoftware** [comment]
I believe this issue is resolved by the answers provided in #223 and the #230 feature request. Feel free to come back or reopen the issue if that's not the case.
